### PR TITLE
virtcontainers : fix shared dir resource remaining

### PR DESCRIPTION
**Pull request by jshachm**
Before this patch shared dir will reamin when sandox
has already removed.

Do clean up shared dirs after all mounts are umounted

Fix: #291

Signed-off-by: Haomin <caihaomin@huawei.com>

### CODE REVIEW

File: virtcontainers/container.go [hunk 1]
```diff
@@ -405,6 +405,14 @@ func (c *Container) unmountHostMounts() error {
 					"error":     err,
 				}).Warn("Could not umount")
 				return err
+			} else if err := os.RemoveAll(m.HostPath); err != nil {
+				// since the mounts related to the shared dir is umounted
+				// we need to remove the host path to avoid resource remaining
+				c.Logger().WithFields(logrus.Fields{
```
Comment (jodh-intel) [inline virtcontainers/container.go]: This log call essentially simulates a call to `WithError()` and although it is simpler than having to call `WithError()`, I think it's safer to explicitly call `WithError()` to protect against it changing its behaviour at some future point.

Since this is based on the log call code above, I'd do something like:

```
logger := c.Logger().WithField("host-path", m.HostPath)

if err := syscall.Unmount(m.HostPath, 0); err != nil {
    logger.WithError(err).Warn("could not unmount")
} else if err := os.RemoveAll(m.HostPath); err != nil {
    logger.WithError(err).Warn("could not remove path")
}
```

File: virtcontainers/container.go [hunk 2]
```diff
@@ -405,6 +405,14 @@ func (c *Container) unmountHostMounts() error {
 					"error":     err,
 				}).Warn("Could not umount")
 				return err
+			} else if err := os.RemoveAll(m.HostPath); err != nil {
```
Comment (jodh-intel) [inline virtcontainers/container.go]: I think you should be able to create a unit test for this in `container_test.go` where you set `c.mounts[0]` to contain a double bind-mounted path. That way, the `Unmount()` will succeed but the `RemoveAll()` will fail (as there is still one more `Unmount()` call to make).

File: virtcontainers/container.go (hunk 1 above)
Comment (jshachm) [reply virtcontainers/container.go]: Going to change it now~

File: virtcontainers/sandbox.go [hunk 3]
```diff
@@ -1162,6 +1162,12 @@ func (s *Sandbox) stop() error {
 		return err
 	}
 
+	// vm is stopped remove the sandbox shared dir
+	sandboxSharedDir := filepath.Join(kataHostSharedDir, s.id)
```
Comment (sboeuf) [inline virtcontainers/sandbox.go]: This cannot be done here. `kataHostSharedDir` is specific to the `kata_agent.go` implementation, and should never be used by the generic sandbox implementation.

Comment (jshachm) [reply virtcontainers/sandbox.go]: Sry for missing this. This should be done more generic.

**Comment by jshachm** [comment]
Need to find a better way to remove sandbox work dirs and solve the case that shared dirs are mounted to different destinations. So do-not-merge by now

File: virtcontainers/sandbox.go (hunk 3 above)
Comment (amshinde) [reply virtcontainers/sandbox.go]: @jshachm You can introduce an agent interface to cleanup agent specific artifacts after the VM has stopped.

**Comment by jshachm** [comment]
@amshinde Good suggestions. Working on it and maybe it's good for #351

File: virtcontainers/sandbox.go (hunk 3 above)
Comment (sboeuf) [reply virtcontainers/sandbox.go]: But wait, any cleanup specific to an agent implementation should be part of either `stopContainer()` if it's related to a container, or `stopSandbox()` if it's related to the sandbox itself.
I don't see the value (and actually this would imply duplication) of introducing a new method to the agent interface.

Comment (amshinde) [reply virtcontainers/sandbox.go]: @sboeuf The kata host directory can only be deleted after the VM has stopped since it is shared with the VM through 9p. So this cannot be part of `stopContainer` or `stopSandbox`.

Comment (sboeuf) [reply virtcontainers/sandbox.go]: Oh yes good point, then in this case, a `agent.cleanup()` function as you suggested sounds like a good idea.

**Comment by katabuilder** [comment]
PSS Measurement:
Qemu: 159779 KB
Proxy: 8730 KB
Shim: 10810 KB

Memory inside container:
Total Memory: 2045972 KB
Free Memory:  1996996 KB

**Comment by jshachm** [comment]
@sboeuf @amshinde    get `cleanupSandbox` done.

Maybe need more unit test as @jodh-intel suggested. Will pull those later.

**Comment by amshinde** [comment]
thanks @jshachm!
lgtm

File: virtcontainers/kata_agent.go [hunk 4]
```diff
@@ -528,6 +528,14 @@ func (k *kataAgent) stopSandbox(sandbox *Sandbox) error {
 	return k.proxy.stop(sandbox, k.state.ProxyPid)
 }
 
+func (k *kataAgent) cleanupSandbox(sandbox *Sandbox) error {
+	sandboxSharedDir := filepath.Join(kataHostSharedDir, sandbox.id)
+	if err := os.RemoveAll(sandboxSharedDir); err != nil {
```
Comment (sboeuf) [inline virtcontainers/kata_agent.go]: Please simplify the function this way:
```go
func (k *kataAgent) cleanupSandbox(sandbox *Sandbox) error {
        return os.RemoveAll(filepath.Join(kataHostSharedDir, sandbox.id))
}
```

File: virtcontainers/sandbox.go [hunk 5]
```diff
@@ -1181,6 +1181,12 @@ func (s *Sandbox) stop() error {
 		return err
 	}
 
+	// vm is stopped remove the sandbox shared dir
+	if err := s.agent.cleanupSandbox(s); err != nil {
+		// cleanup resource failed shouldn't block destroy sandbox
+		// just raise a warning
+		s.Logger().WithError(err).Warnf("cleanup sandbox failed")
```
Comment (sboeuf) [inline virtcontainers/sandbox.go]: Looks fair !

File: virtcontainers/sandbox.go [hunk 6]
```diff
@@ -1181,6 +1181,12 @@ func (s *Sandbox) stop() error {
...
+	}
```
Comment (sboeuf) [inline virtcontainers/sandbox.go]: Please add a blank line to let the code breathe :)

File: virtcontainers/virtcontainers_test.go [hunk 7]
```diff
@@ -41,7 +41,7 @@ var testQemuPath = ""
 var testHyperstartCtlSocket = ""
 var testHyperstartTtySocket = ""
 
-// cleanUp Removes any stale sandbox/container state that can affect
+// cleanup Removes any stale sandbox/container state that can affect
```
Comment (sboeuf) [inline virtcontainers/virtcontainers_test.go]: This should not be part of this PR.

**Comment by jshachm** [comment]
@sboeuf all changes done~

**Comment by sboeuf** [comment]
Thanks I'll take a look !

**Comment by sboeuf** [comment]
@jshachm LGTM
Could you please repush to retrigger the CI now that we should not have glitches anymore.

File: virtcontainers/container.go [hunk 8]
```diff
@@ -510,11 +510,14 @@ func (c *Container) mountSharedDirMounts(hostSharedDir, guestSharedDir string) (
...
 	for _, m := range c.mounts {
 		if m.HostPath != "" {
+			logger := c.Logger().WithField("host-path", m.HostPath)
 			if err := syscall.Unmount(m.HostPath, 0); err != nil {
-				c.Logger().WithFields(logrus.Fields{
-					"host-path": m.HostPath,
-					"error":     err,
-				}).Warn("Could not umount")
+				logger.WithError(err).Warn("Could not umount")
+				return err
```
Comment (grahamwhaley) [inline virtcontainers/container.go]: Not a change by your patch, but directly related - it seems if we fail one unmount then we quit the loop. Would it maybe be better for us to continue to try and unmount the remaining mounts in `c.mounts`, and clean up as much as we can (even though we know we will ultimately fail this function)?
Same for the following `RemoveAll` as well.

**Comment by jodh-intel** [comment]
The test failures I think are https://github.com/kata-containers/tests/issues/427 (see https://github.com/kata-containers/runtime/pull/385#issuecomment-397922893).

File: virtcontainers/container.go (hunk 8 above)
Comment (jshachm) [reply virtcontainers/container.go]: @grahamwhaley  great catch and suggestions~

When I coded these part, it also took me a lot of time to think whether `continue to next` is better than `break from here`.After thinking and discussion with teammates,  and some "sad" experience from our production environment, I think `umount` failure is a very bad situation maybe caused by `mount point leak` and other bad thing... So we need to break right here and raise a warning to upper platform like `k8s`. 

In other word, If a remaining has already happened, maybe it better to alarm the `SRE` department as quick as possible to get involved in this problem.

What do you think about this, glad to hear your suggestion in this situation~~~ ^_^

**Comment by jshachm** [comment]
@grahamwhaley thx for the remind of `commit msg`. I will update it and make it more clear once I get to my workstation next morning... And for CI thing, with the help of @chavafg I finally get my button back~ 

@jodh-intel As you mentioned , CI will not be happy until we change the scripts of ping CNI?

File: virtcontainers/container.go (hunk 8 above)
Comment (grahamwhaley) [reply virtcontainers/container.go]: Hi @jshachm Yeah, it is a tricky call if we quit or carry on. I can see your point that 'fail fast' and try not to make things worse is perfectly valid. Maybe that is the safer way which will:
- possibly cause 'less damage' if things are really broken
- give the admins more of a chance to diagnose the problem

So, I'm happy to leave as is. If we do, can we add a big comment saying why we quit on the first failure ;-)
Probably best if we get some other input from maybe @egernst and @gnawux maybe, and others if anybody has a strong opinion on this?

**Comment by sboeuf** [comment]
@jshachm if you rebase your PR on top of master (make sure you fetch), you should get the CI test passing since the CNI issue got fixed.

**Comment by amshinde** [comment]
@jshachm Yes, CI has been fixed now. Please rebase on top of master.
